server/engine/src/main/java/org/eclipse/lsp/cobol/lsp/handlers/extended/AnalysisHandler.java

```diff
@@ -36,70 +43,92 @@
+import org.eclipse.lsp4j.Position;
 
 /**
  * Our Extension to LSP Analysis Handler
  */
 @Slf4j
 public class AnalysisHandler {
-  private final AsyncAnalysisService asyncAnalysisService;
-  private final AnalysisService analysisService;
-  private final CFASTBuilder cfastBuilder;
-  private final Communications communications;
-  private final DocumentModelService documentModelService;
+    private final AsyncAnalysisService asyncAnalysisService;
+    private final AnalysisService analysisService;
+    private final CFASTBuilder cfastBuilder;
+    private final Communications communications;
+    private final DocumentModelService documentModelService;
+
+    @Inject
+    public AnalysisHandler(AsyncAnalysisService asyncAnalysisService, AnalysisService analysisService, CFASTBuilder cfastBuilder, Communications communications, DocumentModelService documentModelService) {
+        this.asyncAnalysisService = asyncAnalysisService;
+        this.analysisService = analysisService;
+        this.cfastBuilder = cfastBuilder;
+        this.communications = communications;
+        this.documentModelService = documentModelService;
+    }
 
-  @Inject
-  public AnalysisHandler(AsyncAnalysisService asyncAnalysisService, AnalysisService analysisService, CFASTBuilder cfastBuilder, Communications communications, DocumentModelService documentModelService) {
-    this.asyncAnalysisService = asyncAnalysisService;
-    this.analysisService = analysisService;
-    this.cfastBuilder = cfastBuilder;
-    this.communications = communications;
-    this.documentModelService = documentModelService;
-  }
+    /**
+     * Handle analysis request from our LSP extension
+     *
+     * @param analysisResultEvent request params
+     * @return ExtendedApiResult
+     * @throws ExecutionException   forward the exception
+     * @throws InterruptedException forward the exception
+     */
+    public ExtendedApiResult analysis(AnalysisResultEvent analysisResultEvent) throws ExecutionException, InterruptedException {
+        String uri = analysisResultEvent.getUri();
+        CobolDocumentModel doc = documentModelService.get(uri);
+        if (analysisService.isCopybook(uri, analysisResultEvent.getText())) {
+            communications.notifyGeneralMessage(MessageType.Info, "Cannot retrieve outline tree because file was treated as a copybook");
+            return new ExtendedApiResult(ImmutableList.of());
+        }
+        RootNode rootNode = doc.getLastAnalysisResult().getRootNode();
+        Position position = new Position(analysisResultEvent.getLine(), analysisResultEvent.getCharacter());
+        Optional<Node> selectedNode = RangeUtils.findNodeByPosition(rootNode, analysisResultEvent.getUri(), position);
+        if (selectedNode.isPresent() && !(selectedNode.get() instanceof RootNode)) {
+            return cfastBuilder.build((ProgramNode) selectedNode.get()
+                    .getNearestParentByType(PROGRAM).orElse(rootNode.findFirstProgramNode()));
+        }
```

Comment:
is this logic better in some way compared to the `selectProgramNode` approach below?

Reply:
A bit, findNodeByPosition also consider copybooks/uri in the search.